### ISSUE TITLE
Segmented DA

### ISSUE THREAD

**Issue by gavofyork** [gavofyork/graypaper#1]
Target: 2048 segments import/extrinsics and export/extrinsic per WP, each 4096 byte

- Unfinished
  - [ ] Limit number of exports, not just size.
  - [ ] Ensure the 2048 is a limit for total manifest entries (import, extrinsic, export).
  - [ ] Remove the regular hash from manifests.
  - [ ] In doing so will need to properly consider how to reconstruct the WP for auditing.

- Data
  - Work Package has extrinsics section including hash of all extrinsic data. Extrinsics are distributed with it.
  - Work Package has imports section including:
    - ECroot and segment index
    - Q: What about segment offset/length? Is it only in WP? What if we want to recompute from master?
  - [ ] `export` host-call for Refine
  - [ ] `imports` argument for Refine
  - [ ] `imports` in Work Package
  - [ ] `extrinsics` in Work Package

- [ ] Distribute provision in single blob EC'd and entered into DA, max 15MB message length
  - [ ] Quantize & zero-pad segments to whatever makes sense in DA (2KB currently)
  - [ ] Segments: [WP ++ Extrinsic_0 ++ Extrinsic_1 ++ ... Export_0 ++ Export_1 ++ ...]
  - [ ] Assurers identify provision through EC-root
    - Q: identify also through size?
  - [ ] EC-root is M2 root of seq of pairs, one for each chunk:
    - [ ] Master chunk hash
    - [ ] M2 root of each segment chunk hash, one for each segment

- [ ] Guarantor task
  - Guarantors calculate WR and distribute master chunks in usual way.
  - Guarantors calculate EC-root, including the extra segment-chunk merklisation.
- [ ] Guarantors must be able to have confidence that proofs are good
  - Q: Is it realistic to get all import data with proofs from WP author?
    - Segment proof size = (10 + 11) * 32 = 672bytes per validator, 340 validators = 220KB/seg
    - 2048 segments x 220KB = 457MB
    - No.
  - Guarantors optimistically ask for imports from randomly selected validators.
  - Honest ones always do random sampling to avoid poisoned import data.
  - They keep all signed import data witnesses as these can transfer blame in the case of a slash.
  - Unlikely that two honest guarantors will end up with same invalid WR.
  - If one is dishonest, they could go along with bad WR, but then honest would be able to transfer blame and dishonest would get slashed.
  - [ ] Include manifest distribution in bandwidth: 511 * 2048 * 36 = 36 MB / 6s = 6 MB/s = 48Mbps
  - [ ] Ensure length of imports + WP <= 15MB, imports_count <= 2048
  - [ ] Fetch/reconstruct all Imports
    - Optimistically from random selection of 340 peers.
    - Require signed, personalised statements in case of need to prove fraud later.
    - Q: Worth adding more erasure-coding to be able to detect bad chunks?
  - After execution:
  - [ ] Disregard `export`s which would take length exports + WP > 15MB or exports_count > 2048
  - [ ] Erasure-code [WP ++ Extrinsics ++ Exports]
  - [ ] Distribute chunks
  - [ ] Construct and distribute `import_manifest` all ECroot+index for each import (2048 * 36bytes = 72KB * 511 = 36MB -> 6MB/s -> 48Mbps)
  - [ ] Merklize import_manifest to create WR `import_manifest_root`
- [x] Assurer task
  - [x] Check receipt of WP and exports, as well as prescience of imports preimages, allowing computation of WR's `manifest_root` from first-principles.
- [ ] Properly detail auditor verification task
  - [ ] Fetch/reconstuct WP
  - [ ] Ensure length of imports + extrinsics + WP <= 15MB (assurer-upload/auditor-download)
  - [ ] Ensure |imports| + |extrinsics| + |exports| <= 2048 (manifest distribution)
  - [ ] Fetch/reconstruct imports
    - Same as for guarantor
  - [ ] Fetch/reconstruct extrinsics
    - Harder since WP contains extrinsic hash, but assurers only know by erasure-root.
    - Option 1: download the manifest and cross-reference - an extra ~136KB.
    - Option 2: include erasure roots in the WP's extrinsic spec - extra work for author
  - [ ] Erasure-code all extrinsics
  - After execution:
  - [ ] Ensure length of WP + extrinsics + exports <= 15MB (guarantor-upload/assurer-download)
  - [ ] Disregard `export`s which would take total length exports + WP > 15MB or exports_count > 2048
  - [ ] Erasure-code each export
  - [ ] Merklize WP imports field to verify WR `import_manifest_root`
  - Q: Still need this? Merklize erasure-coded {extrinsics, exports} to verify WR `providers_root`

TODO:
- [ ] Look into using exotic crypto to encode a content verifier and allow for import data hashes to be validated succinctly:
  - basically we want a primitive which takes `[H0, H1, H2, ..., H2047]` and gives some commitment `C` back which is much much less than 64KB
  - and then there's some function `f: (C, i, H) -> bool` which returns true iff `Hi = H`, but unlike a Merkle root does so without the need for supplying any witness data (since we don't want to store all of those hashes)

**Comment by seunlanlege** [comment]
> Look into using exotic crypto to encode a content verifier and allow for import data hashes to be validated succinctly:
basically we want a primitive which takes [H0, H1, H2, ..., H2047] and gives some commitment C back which is much much less than 64KB
and then there's some function f: (C, i, H) -> bool which returns true iff Hi = H, but unlike a Merkle root does so without the need for supplying any witness data (since we don't want to store all of those hashes)

I believe this can be achieved with a KZG commitment? Commitment `C` depending on your choice of elliptic curves will be at most 96bytes. It does need a proof/witness, but luckily proof sizes are constant-sized at a maximum of 96bytes as well.

**Comment by gavofyork** [comment]
> unlike a Merkle root does so without the need for supplying **any** witness data

Unfortunately KZG isn't going to work as it still requires some witness data.
